datastream-kafka/src/main/java/com/linkedin/datastream/kafka/KafkaProducerWrapper.java

```diff
@@ -254,67 +298,92 @@ void send(DatastreamTask task, ProducerRecord<K, V> producerRecord, Callback onC
-    // quickly to avoid delaying/blocking the sender thread. Thus schedule the actual producer.close() on a separate
-    // thread
-    if (producer != null) {
+      // This may be called from the send callback. The callbacks are called from the sender thread, and must complete
+      // quickly to avoid delaying/blocking the sender thread. Thus schedule the actual producer.close() on a separate
+      // thread
+
       _producerCloseExecutorService.submit(() -> {
         _log.info("KafkaProducerWrapper: Closing the Kafka Producer");
-        producer.close(CLOSE_TIMEOUT_MS, TimeUnit.MILLISECONDS);
-        NUM_PRODUCERS.decrementAndGet();
-        _log.info("KafkaProducerWrapper: Kafka Producer is closed");
+        try {
+          producer.close(CLOSE_TIMEOUT_MS, TimeUnit.MILLISECONDS);
+          NUM_PRODUCERS.decrementAndGet();
+          _log.info("KafkaProducerWrapper: Kafka Producer is closed");
+        } finally {
+          try {
+            _producerLock.lock();
+            closeInProgress = false;
+            _waitOnNoProducerClose.signalAll();
+          } finally {
+            _producerLock.unlock();
+          }
+        }
       });
+    } finally {
+      _producerLock.unlock();
     }
   }
 
-  private DatastreamRuntimeException generateSendFailure(Exception exception) {
+  private DatastreamRuntimeException generateSendFailure(Exception exception, DatastreamTask task) {
     _dynamicMetricsManager.createOrUpdateMeter(_metricsNamesPrefix, AGGREGATE, PRODUCER_ERROR, 1);
     if (exception instanceof IllegalStateException) {
       _log.warn("Send failed transiently with exception: ", exception);
       return new DatastreamTransientException(exception);
     } else {
       _log.warn("Send failed with a non-transient exception. Shutting down producer, exception: ", exception);
-      shutdownProducer();
+      if (_tasks.contains(task)) {
+        shutdownProducer();
+      }
       return new DatastreamRuntimeException(exception);
     }
   }
 
   void flush() {
-    synchronized (_producerLock) {
-      try {
-        if (_kafkaProducer != null) {
-          _kafkaProducer.flush(_producerFlushTimeoutMs, TimeUnit.MILLISECONDS);
-        }
-      } catch (InterruptException | TimeoutException e) {
-        // The KafkaProducer object should not be reused on an interrupted flush
-        _log.warn("Kafka producer flush interrupted/timed out, closing producer {}.", _kafkaProducer);
+    Producer<K, V> producer;
+    try {
+      _producerLock.lock();
+      producer = _kafkaProducer;
+    } finally {
+      _producerLock.unlock();
+    }
+
+    try {
+      if (producer != null) {
+          producer.flush(_producerFlushTimeoutMs, TimeUnit.MILLISECONDS);
+      }
+    } catch (InterruptException | TimeoutException e) {
+      // The KafkaProducer object should not be reused on an interrupted flush
+      if (producer == _kafkaProducer) {
```

Comment:
The reason to skip here is to avoid closing the other producer. I initially thought of passing producer to the method, but that will imply making the producer final to be passed from send callback and I wanted to avoid it from garbage collection point of view and producer is a heavyweight object in general.

Reply:
I wasn't implying always passing the `producer` object to `shutdownProducer()`. I was thinking that you can keep `shutdownProducer()` as is, but add:

`
void shutdownProducer(KafkaProducer producer) {
    _producerLock.lock()
    try {
        if (producer == _kafkaProducer) {
             shutdownProducer()
       }
    } finally {
        _producerLock.unlock()
    }
}
`

This new method would be called from flush() only. The send callback path will just call the existing `shutdownProducer()` directly, since this method acquires the lock anyways.

I'd like to avoid access to `_kafkaProducer` outside the lock as much as possible. Makes the code much cleaner and easier to reason about. Otherwise one will always need to remember that another code path can potentially have changed the  `_kafkaProducer` object from underneath us by the time we actually call `shutdownProducer()`, implying that we've just closed a newly created producer, etc.